File: UniversityApi.API/Repositories/GenericRepository.cs
using Microsoft.EntityFrameworkCore;
using UniversityApi.API.Contracts;
using UniversityApi.API.DataBase;
using UniversityApi.API.DataBase.Entities;

namespace UniversityApi.API.Repositories
{
    public class GenericRepository<T> : IGenericRepository<T> where T : class
    {
        protected readonly UniversityApiDbContext _context;

        public GenericRepository(UniversityApiDbContext dbContext)
        {
            _context = dbContext;
        }

        public virtual async Task<T> AddAsync(T entity)
        {
            var result = await _context.AddAsync(entity);
            await _context.SaveChangesAsync();
            return entity;
        }

        public virtual async Task DeleteAsync(Guid id)
        {
            var entity = await GetAsync(id);
            _context.Set<T>().Remove(entity);
            await _context.SaveChangesAsync();
        }

        public async Task<bool> Exists(Guid id)
        {
            var entity = await GetAsync(id);
            return entity != null;
        }

        public virtual async Task<List<T>> GetAllAsync()
        {
            return await _context.Set<T>().ToListAsync();
        }

        public virtual async Task<T> GetAsync(Guid id)
        {
            return await _context.Set<T>().FindAsync(id);
        }

        public virtual async Task UpdateAsync(T entity)
        {
            _context.Update(entity);
            await _context.SaveChangesAsync();
        }
    }
}

File: FsUniversityApi/Repositories/GenericRepository.fs
namespace FsUniversityApi.Repositories

open Microsoft.EntityFrameworkCore
open FsUniversityApi.Database.FsDbContext
open FsUniversityApi.Database.Entities.ISoftRemovableEntity
open FsUniversityApi.Contracts
open System.Threading.Tasks
open System

type GenericRepository<'T when 'T :> ISoftRemovableEntity and 'T : not struct>(context : FsDbContext) =
    interface IGenericRepository<'T> with
        member this.AddAsync (entity : 'T) =
            async {
                do! context.AddAsync(entity).AsTask() |> Async.AwaitTask |> Async.Ignore
                do! context.SaveChangesAsync() |> Async.AwaitTask |> Async.Ignore
                return entity
            }
            |> Async.StartAsTask

        member this.DeleteAsync (id : Guid) =
            async {
                let! (entity : ISoftRemovableEntity) = context.FindAsync(id).AsTask() |> Async.AwaitTask
                let updatedEntity = entity.SetSoftDeleted(true)
                context.Remove(entity) |> ignore
                do! context.AddAsync(entity).AsTask() |> Async.AwaitTask |> Async.Ignore
                return! context.SaveChangesAsync() |> Async.AwaitTask
            }
            |> Async.StartAsTask :> Task

        member this.Exists (id : Guid) =
            async {
                let! entity = context.FindAsync(id).AsTask() |> Async.AwaitTask
                return entity <> null
            }
            |> Async.StartAsTask

        member this.GetAllAsync () =
            context.Set<'T>().ToListAsync()

        member this.GetAsync (id : Guid) =
            context.Set<'T>().FindAsync(id).AsTask()

        member this.UpdateAsync (entity : 'T) =
            async {
                let! entity = context.FindAsync(id).AsTask() |> Async.AwaitTask
                context.Remove(entity) |> ignore
                do! context.AddAsync(entity).AsTask() |> Async.AwaitTask |> Async.Ignore
                return! context.SaveChangesAsync() |> Async.AwaitTask
            }
            |> Async.StartAsTask :> Task